## Task 3: Data Analysis

2011_Transactions.csv is a database of transactions in the travel industry. This task requires you to provide summary statistics and run some basic analysis on this file

In [1]:
import pandas as pd 

In [2]:
# read dataset as pandas dataframe 

trans = pd.read_csv('2011_Transactions.csv', encoding= 'unicode_escape')

In [3]:
# check column information in dataset 

trans.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8437 entries, 0 to 8436
Data columns (total 21 columns):
machine_id            8437 non-null int64
site_session_id       8437 non-null int64
domain_id             8437 non-null uint64
event_date            8437 non-null int64
event_time            8344 non-null object
prod_category_id      8437 non-null int64
prod_name             8437 non-null object
prod_qty              8437 non-null int64
prod_totprice         8437 non-null float64
basket_tot            8437 non-null float64
hoh_most_education    8437 non-null int64
census_region         8437 non-null object
household_size        8437 non-null object
hoh_oldest_age        8437 non-null int64
household_income      8437 non-null int64
children              8437 non-null int64
racial_background     8437 non-null int64
connection_speed      8437 non-null int64
country_of_origin     8437 non-null int64
zip_code              8437 non-null int64
domain_name           8437 non-null object
d

In [4]:
# check the first 5 lines in dataset 

trans.head(5)

machine_id  site_session_id             domain_id  event_date event_time  \
0    13512886    3669584318587   9530952911301729568    20110720    2:03:02   
1    49645796    3355708493805  17475197073474272331    20110228   21:16:51   
2    58622574    3360275042310   1875457313788268580    20110325   15:58:34   
3    59850911    3815456641153   1875457313788268580    20110726   21:31:36   
4    59850911   71716842443028   1875457313788268580    20111220   23:29:16   

   prod_category_id                                          prod_name  \
0                45          CAR - COMPACT CAR JUL 28 2011~JUL 31 2011   
1                43  FLIGHT - COLUMBUS PORT COLUMBUS INTERNATIONAL ...   
2                43                                       FLIGHT - N/A   
3                43                                       FLIGHT - N/A   
4                43                                       FLIGHT - N/A   

   prod_qty  prod_totprice  basket_tot  ...  census_region household_size  \
0         3         147.00      147.00  ...              3              4   
1         1         278.00      278.00  ...              2              3   
2         2         701.96      544.76  ...              3              3   
3         1         451.60      465.60  ...              3              5   
4         1         640.93      662.33  ...              3              5   

  hoh_oldest_age  household_income  children  racial_background  \
0              8                 7         0                  1   
1              9                 2         1                  1   
2              9                 7         0                  1   
3              7                 7         1                  1   
4              7                 7         1                  1   

   connection_speed  country_of_origin  zip_code      domain_name  
0                 1                  0     38016      expedia.com  
1                 1                  0     43224  travelocity.com  
2                 1                  0     32407        delta.com  
3                 1                  0     29909        delta.com  
4                 1                  0     29909        delta.com  

[5 rows x 21 columns]

1. What is the mean number of transactions made per individual (machine_id)? Restricting to the travel categories (prod_category_id = 43, 44 and 45) what is the mean number of transactions?

In [7]:
trans['machine_id'].value_counts().reset_index().mean(axis=0)

index         8.607783e+07
machine_id    2.015047e+00
dtype: float64

The mean number of transactions made per individual (machine_id) is 2.015. 

In [8]:
# Restricting to the travel categories (prod_category_id = 43, 44 and 45) 

travCate = trans[trans['prod_category_id'].isin([43, 44, 45])]

In [9]:
travCate['machine_id'].value_counts().reset_index().mean(axis=0)

index         8.607268e+07
machine_id    1.958835e+00
dtype: float64

The mean number of transactions made per individual (machine_id) with travel categories restriction is 1.959.

2. Restricting attention again to the travel categories, for those who had multiple transactions:
 - How far apart are the transactions? Using this, categorize transactions that are likely related to each other, e.g. car rented and flight booked for the same trip.
 - For transactions related to each other, document the sequence of bookings (e.g. car booked first, then air travel, then hotel) in a visual manner.

In [7]:
travCate

machine_id  site_session_id             domain_id  event_date  \
0       13512886    3669584318587   9530952911301729568    20110720   
1       49645796    3355708493805  17475197073474272331    20110228   
2       58622574    3360275042310   1875457313788268580    20110325   
3       59850911    3815456641153   1875457313788268580    20110726   
4       59850911   71716842443028   1875457313788268580    20111220   
...          ...              ...                   ...         ...   
8432    95319234    7671937175476  14244539005989555924    20110102   
8433    95319894    5598578479186   5127767531286671227    20110609   
8434    95319894    5598578479186   5127767531286671227    20110609   
8435    95319894   68167160172789  14244539005989555924    20111119   
8436    95320067    5474873184413   3010609366849421442    20110823   

     event_time  prod_category_id  \
0       2:03:02                45   
1      21:16:51                43   
2      15:58:34                43   
3      21:31:36                43   
4      23:29:16                43   
...         ...               ...   
8432    4:16:06                43   
8433    0:21:18                43   
8434    0:21:18                43   
8435   20:40:14                43   
8436    0:07:21                45   

                                              prod_name  prod_qty  \
0             CAR - COMPACT CAR JUL 28 2011~JUL 31 2011         3   
1     FLIGHT - COLUMBUS PORT COLUMBUS INTERNATIONAL ...         1   
2                                          FLIGHT - N/A         2   
3                                          FLIGHT - N/A         1   
4                                          FLIGHT - N/A         1   
...                                                 ...       ...   
8432  FLIGHT -WASHINGTON, DC (DCA)~AUSTIN, TX (AUS) N/A         1   
8433                       FLIGHT TUESDAY, JULY 5, 2011         1   
8434                      FLIGHT SATURDAY, JULY 2, 2011         1   
8435     FLIGHT -HOUSTON, TX (IAH)~DENVER, CO (DEN) N/A         1   
8436  NEW YORK CITY, NY (JFK)~DENVER, CO (DEN) FRI A...         1   

      prod_totprice  basket_tot  ...  census_region household_size  \
0            147.00      147.00  ...              3              4   
1            278.00      278.00  ...              2              3   
2            701.96      544.76  ...              3              3   
3            451.60      465.60  ...              3              5   
4            640.93      662.33  ...              3              5   
...             ...         ...  ...            ...            ...   
8432         111.65      111.65  ...              1              2   
8433         434.00      434.00  ...              1              1   
8434           0.00      434.00  ...              1              1   
8435         137.90      137.90  ...              1              1   
8436         508.00      508.00  ...              4              4   

     hoh_oldest_age  household_income  children  racial_background  \
0                 8                 7         0                  1   
1                 9                 2         1                  1   
2                 9                 7         0                  1   
3                 7                 7         1                  1   
4                 7                 7         1                  1   
...             ...               ...       ...                ...   
8432              3                 1         0                  1   
8433              5                 1         0                  3   
8434              5                 1         0                  3   
8435              5                 1         0                  3   
8436              7                 7         1                  1   

      connection_speed  country_of_origin  zip_code      domain_name  
0                    1                  0     38016      expedia.com  
1                    1                  0     43224  tr

In [26]:
# The transactions with same machine_id appearing at least twice in the dataset 

travCate['machine_id'].value_counts().loc[lambda x : x > 1] 

93855448    34
85432156    29
82495406    28
91433731    26
81202618    24
            ..
83749430     2
89076285     2
90022106     2
80261697     2
87107610     2
Name: machine_id, Length: 1632, dtype: int64

In [9]:
# check the transactions with a specific machine_id with sorted dates 

travCate[travCate['machine_id'] == 82495406].sort_values(by=['event_date', 'event_time'])

machine_id  site_session_id             domain_id  event_date  \
2355    82495406    4307935170488   1037677886457106237    20110106   
2358    82495406    4340104368144   1037677886457106237    20110404   
2359    82495406    4340104368144   1037677886457106237    20110404   
2360    82495406    4340104368144   1037677886457106237    20110404   
2361    82495406    4340104368144   1037677886457106237    20110404   
2349    82495406    4222094741522   1037677886457106237    20110406   
2350    82495406    4222094741522   1037677886457106237    20110406   
2351    82495406    4222094741522   1037677886457106237    20110406   
2354    82495406    4295779160166  10326615452578885078    20110629   
2348    82495406    4216806969452   1037677886457106237    20110705   
2353    82495406    4284431208557  10326615452578885078    20110706   
2357    82495406    4315771048053   1037677886457106237    20110714   
2356    82495406    4314067374201   1037677886457106237    20110718   
2352    82495406    4273007235202   7101213156062330967    20110727   
2362    82495406    4452757016732  10326615452578885078    20110822   
2364    82495406   69766386684102   1037677886457106237    20111003   
2365    82495406   69766386684102   1037677886457106237    20111003   
2366    82495406   69766386684102   1037677886457106237    20111003   
2367    82495406   69911374860488   1037677886457106237    20111005   
2368    82495406   69911374860488   1037677886457106237    20111005   
2369    82495406   69911374860488   1037677886457106237    20111005   
2371    82495406   71805965832400  10326615452578885078    20111013   
2370    82495406   70951134695652  10326615452578885078    20111102   
2375    82495406   73582287655154   1037677886457106237    20111116   
2363    82495406   69682984784130  10326615452578885078    20111202   
2372    82495406   72042780168468   1037677886457106237    20111220   
2373    82495406   72042780168468   1037677886457106237    20111220   
2374    82495406   72042780168468   1037677886457106237    20111220   

     event_time  prod_category_id  \
2355   17:55:12                43   
2358   16:50:50                43   
2359   16:50:50                43   
2360   16:50:50                45   
2361   16:50:50                43   
2349   17:48:44                43   
2350   17:48:44                45   
2351   17:48:44                43   
2354   14:23:12                43   
2348   15:56:38                43   
2353   14:56:57                43   
2357   18:25:41                43   
2356   17:03:00                43   
2352   15:34:30                43   
2362   16:45:04                43   
2364   13:28:54                43   
2365   13:28:54                45   
2366   13:28:54                43   
2367   14:28:43                43   
2368   14:28:43                43   
2369   14:28:43                45   
2371   15:00:27                43   
2370   16:50:27                43   
2375   21:24:32                43   
2363   22:14:09                43   
2372   17:32:31                43   
2373   17:32:31                45   
2374   17:32:31                43   

                                              prod_name  prod_qty  \
2355                              FLIGHT - ADULT JAN~13         1   
2358                               FLIGHT - ADULT MAY~2         1   
2359                               FLIGHT - ADULT MAY~2         1   
2360  CAR - ECONOMY CAR MONDAY, MAY 2, ~WEDNESDAY, M...         2   
2361                              UNLIMITED MILEAGE N/A         1   
2349                              FLIGHT - ADULT APR~27         2   
2350  CAR - ECONOMY CAR WEDNESDAY, APRIL 27, ~WEDNES...         1   
2351                              UNLIMITED MILEAGE N/A         1   
2354                                 ADULT JUL 12, 2011         1   
2348                                 FLIGHT - ADULT N/A         1   
2353                                 ADULT JUL 13, 2011         1   
2357                                 FLIGHT 

In [30]:
travCate[(travCate['machine_id'] == 93855448) & (travCate['event_date'] == 20110917)]

machine_id  site_session_id            domain_id  event_date event_time  \
7344    93855448    5833288192182  1037677886457106237    20110917   12:26:20   
7345    93855448    5833288192182  1037677886457106237    20110917   12:26:20   
7346    93855448    5833288192182  1037677886457106237    20110917   12:26:20   
7347    93855448    5833288192182  1037677886457106237    20110917   12:26:20   
7348    93855448    5833288192182  1037677886457106237    20110917   12:26:20   
7349    93855448    5833288192182  1037677886457106237    20110917   12:26:20   
7350    93855448    5833288192182  1037677886457106237    20110917   12:26:20   
7351    93855448    5833288192182  1037677886457106237    20110917   12:26:20   
7352    93855448    5833288192182  1037677886457106237    20110917   12:26:20   
7353    93855448    5833288192182  1037677886457106237    20110917   12:26:20   
7354    93855448    5833288257718  1037677886457106237    20110917   13:17:39   
7355    93855448    5833288257718  1037677886457106237    20110917   13:17:39   

      prod_category_id                   prod_name  prod_qty  prod_totprice  \
7344                43      FLIGHT - AWARD TPA-SDF         1           2.50   
7345                43      FLIGHT - AWARD SDF-TPA         1           2.50   
7346                43                FLIGHT - N/A         1           5.01   
7347                43  FLIGHT - COMPANION SDF-TPA         1           2.50   
7348                43  FLIGHT - COMPANION TPA-SDF         1           2.50   
7349                43                FLIGHT - N/A         1           0.00   
7350                43                FLIGHT - N/A         1           2.50   
7351                43  FLIGHT - COMPANION TPA-SDF         1           2.50   
7352                43  FLIGHT - COMPANION SDF-TPA         1           2.50   
7353                43                FLIGHT - N/A         1           5.01   
7354                43     FLIGHT - POINTS SDF-TPA         1           2.50   
7355                43     FLIGHT - POINTS TPA-SDF         1           2.50   

      basket_tot  ...  census_region household_size hoh_oldest_age  \
7344       25.00  ...              2              3              7   
7345       25.00  ...              2              3              7   
7346       30.01  ...              2              3              7   
7347       30.01  ...              2              3              7   
7348       30.01  ...              2              3              7   
7349       10.00  ...              2              3              7   
7350       12.50  ...              2              3              7   
7351       30.01  ...              2              3              7   
7352       30.01  ...              2              3              7   
7353       30.01  ...              2              3              7   
7354       25.00  ...              2              3              7   
7355       25.00  ...              2              3              7   

      household_income  children  racial_background  connection_speed  \
7344                 5         0                  1                 1   
7345                 5         0                  1                 1   
7346                 5         0                  1                 1   
7347                 5         0                  1                 1   
7348                 5         0                  1                 1   
7349                 5         0                  1                 1   
7350                 5         0                  1                 1   
7351                 5         0                  1                 1   
7352                 5         0                  1                 1   
7353                 5         0                  1                 1   
7354                 5         0                  1                 1   
7355                 5         0                  1                 1   

      country_of_origin  zip_code    domain_name  
7344        